torchtitan/components/checkpoint.py

```diff
@@ -336,35 +369,44 @@ def save(self, curr_step: int, force: bool = False) -> None:
             None
         """
 
+        if self.ft_manager:
+            self._ft_save(curr_step)
+
         if not self._should_save(curr_step, force):
             return
 
         begin = time.monotonic()
-        logger.info("Saving the checkpoint (or staging if async is enabled).")
-        checkpoint_id = self._create_checkpoint_id(curr_step)
-        self._async_wait()
-        # This GC is called for async checkpoint as it is useless to do
-        # GC right after async_save -- the CPU memory is not able to be
-        # freed until _async_wait()
-        if force:
-            self._save_last_step(curr_step)
-        elif self.async_mode == AsyncMode.ASYNC_WITH_PINNED_MEM:
-            GarbageCollection.collect("GC collection invoked by checkpointer.")
-            self._async_with_pinned_memory(checkpoint_id)
-        elif self.async_mode == AsyncMode.ASYNC:
-            GarbageCollection.collect("GC collection invoked by checkpointer.")
-            self.async_future = dcp.async_save(
-                self.states, checkpoint_id=checkpoint_id, process_group=self.pg
-            )
-            GarbageCollection.collect("GC collection invoked by checkpointer.")
-        else:
-            save_with_gc(self.states, checkpoint_id=checkpoint_id)
-        self._purge_stale_checkpoints()
+        if not self.ft_manager or self.ft_manager.participating_rank() == 0:
```

Comment:
This uses `.participating_rank()` but the load logic below uses `ft_replica_id`. Seems like individual replicas may miss the checkpoint file on recovery?

Reply:
There are two save and loads. 

One is full persistent checkpoint, the full checkpoint (model, optimizer, ...). The logic of this is replica that has `participating_rank() == 0` perform `save()` but all replica will load the `full persistent checkpoint` even if the replica is recovering. The `load` can be avoid if the replica is recovering and there is another live replica. But we have not done this optimization.

Another one is ft checkpoint. This only saves per-replica dataloader information. The logic of this every replica saves and loads its own dataloader information to/from different folders -- prefixing with `ft_replica_id`.

Reply:
I added comments to the Checkpointer.